### ISSUE TITLE
Non-grouped aggregation variables

### ISSUE THREAD

**Issue by rnewman** [mozilla/mentat#486]
Datomic enforces that every entry in a `:find` element list is either an aggregate function or a variable _used for grouping_. That is, a query like

```edn
[:find ?x (max ?y) ?z
 :with ?w
 :where …]
```

is equivalent to the SQL:

```sql
SELECT `?x`, max(`?y`), `?z`
FROM …
WHERE …
GROUP BY `?x`, `?z`, `?w`
```

There is _no way_ to get the equivalent of SQL without the `GROUP BY` — the only way to achieve that is to use a subquery:

```clojure
(datomic.api/q
   '[:find ?title ?year
      :in $
      :with ?title
      :where
      [(datomic.api/q '[:find (max ?year)
                        :where [_ :movie/release-year ?year]]
                      $)
       [[?year]]]
      [?x :movie/title ?title]
      [?x :movie/release-year ?year]])
```

or to run two queries.

This is acceptable for 'classic' Datomic, because all queries run on the peer, and can thus intermingle code and queries without issue.

For client-server Datomic and for Mentat this is more problematic: maintaining isolation across two queries or over a network link is inconvenient and harms throughput.

I propose two options:

1. Support subqueries without resorting to nested Clojure API calls.
2. Support [SQLite's guaranteed same-row aggregate values](http://www.sqlite.org/releaselog/3_7_11.html) (or, more generally, returning a single contributing value for an aggregate), via a fake aggregate function, perhaps called `the`.

These two together might look something like:

```edn
[:find (the ?id) (the ?title) (max ?year)
 :where
 [[:find [?actor ...]]]
 [?x :movie/actor ?actor]
 [?x :movie/title ?title]
 [?x :movie/release-year ?year]]
```

**Comment by ncalexan** [comment]
Fascinating.  Would the following be a workable third option?

3. Make `:with` authoritative rather than additive.  

From http://docs.datomic.com/query.html#sec-5-17-1: 

> The solution to this problem is the :with clause, which considers additional variables when forming the basis set for the query result. The :with variables are then removed, leaving a bag (not a set!) of values available for aggregation.

Suppose we said that "the :with clause specifies the complete set of variables when forming the basis set for the query result."  With this interpretation, `:with` translates almost directly to `GROUP BY`.  That allows for `:with []` (handling the case you say can't currently be expressed in Datomic) and `:with [?x ?y ?w]` (handling the existing behaviour).

What might I be missing?

We can bikeshed the name `:with`, which might better be `:grouping`, if this approach has legs.

**Comment by ncalexan** [comment]
> I propose two options:
> 
>     1. Support subqueries without resorting to nested Clojure API calls.

In general, I'm scared of subqueries.  The CC processing is already extremely subtle: I have a hard time believing we can handle arbitrary recursive CCs robustly.

**Comment by rnewman** [comment]
btw, here's a user question about this that illustrates a good example:

https://stackoverflow.com/questions/23215114/datomic-aggregates-usage

> In general, I'm scared of subqueries. The CC processing is already extremely subtle: I have a hard time believing we can handle arbitrary recursive CCs robustly.

I'm pretty optimistic about that: we already generate nested CCs for `not`, and already produce bindings from computed tables, so this would be putting the two together.

Regardless, I expect that this is something we'll need to support at some point — a subquery is the only correct and general solution to this situation, and is of course broadly useful.

My uncertainty is more about whether there's a consistent and understandable shortcut approach to offering greatest-1-of-group in situations where it's correct to do so: that is, detecting a functional dependency and offering syntax.

Datomic is, from one perspective, doing the safe and correct thing here: making `:with` authoritative is aping SQL, and we know that results in situations without a strict functional dependency are non-deterministic.

**Comment by rnewman** [comment]
I should expand on that last point. I think it's _good_ that Datomic doesn't let you write queries that are ambiguous. Many SQL database systems will also reject ambiguous queries. But SQLite doesn't always, and we ca

**Comment by rnewman** [comment]
I'm not sure what happened there…

**Comment by rnewman** [comment]
Concrete proposal:

For a query containing a **single** _discriminating_ aggregate operation — `max`, `min` — and any number of other aggregates (_e.g._, `avg`) we can easily support a pseudo-aggregate operation that returns corresponding values from associated bindings.

We should error out when more than one discriminating aggregate is used.

If we ever support an engine with different guarantees than SQLite, we'll need to translate this kind of query into a subselect instead.

We should document that all corresponding bindings will come from the same 'row', but that if more than one row contributes the max/min value, that which row is used is not defined.

**Comment by rnewman** [comment]
I implemented this in #584.
